pepr/attack_runner.py

```diff
@@ -0,0 +1,208 @@
+"""
+Attack runner for convenient pentesting with multiple configurations.
+"""
+
+import os
+import logging
+import pickle
+
+import numpy as np
+import tensorflow as tf
+import yaml
+from itertools import groupby
+
+from pepr import report
+from pepr.privacy import mia, gmia
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+
+
+def run_attacks(
+    yaml_path, report_save_path, attack_obj_save_path, functions, pdf=False
+):
+    """
+    Run multtiple attacks configured in a YAML file.
+
+    Parameters
+    ----------
+    yaml_path : str
+        Path to attack configuration file in YAML format.
+    report_save_path : str
+        Path where to save the report files.
+    attack_obj_save_path : str
+        Path where to save the serialized attack objects.
+    pdf : bool
+        If true, the attack runner will build the LaTex report and save an PDF to the
+        save path.
+    functions : dict
+        Dictionary containing the functions which return TensorFlow models. The keys
+        should correspond to the strings in the configuration file.
+
+    Returns
+    -------
+    dict
+        Dictionary with attack alias and the corossponding path to the serialized attack
+        object.
+    """
+
+    attack_object_paths = {}
+
+    # Pickle helper
+    def pickle_attack_obj(attack_obj):
+        with open(
+            f"{attack_obj_save_path}/{attack_obj.attack_alias}.pickle", "wb"
+        ) as f:
+            attack_obj.obj_save_path = attack_obj_save_path
+            pickle.dump(attack_obj, f)
+            attack_object_paths[
+                f"{attack_obj.attack_alias}"
+            ] = f"{attack_obj_save_path}/{attack_obj.attack_alias}.pickle"
+
+    # Parse YAML
+    with open(yaml_path) as f_stream:
+        logger.info(f"Parse attack runner configuration file: {yaml_path}")
+        data = yaml.load(f_stream)
+        attack_obj_save_path
+
+        os.makedirs(attack_obj_save_path, exist_ok=True)
+        os.makedirs(report_save_path + "/fig", exist_ok=True)
+        sections = []
+
+        for i, yaml_attack_pars in enumerate(data["attack_pars"]):
+            attack_type = yaml_attack_pars["attack_type"]
+            logger.info(
+                f"Attack number: {i+1}\n\n"
+                "######################## Attack Run ########################"
+            )
+            if attack_type == "mia":
```

Comment:
Can we make the parameter parsing more generic, such that we do not need a case distinction for each attack type? This way it becomes much more complicated to add new attacks and the Attack Runner becomes increasingly complex. Probably we have to update our Attacks a little, but I think that would be worth it :)

Reply:
Thank you for the suggestion! The attack runner was reworked and is way more generic now. Support an new attack is as easy as adding a new entry to a dictionary :partying_face:

Reply:
Looks great 👍